From a0a59d36586086a588215f3f4ffa1734e3a2e10a Mon Sep 17 00:00:00 2001
From: Aaron Plattner <aplattner@nvidia.com>
Date: Thu, 27 May 2021 16:58:56 -0700
Subject: [PATCH 011/105] os: print <signal handler called> if
 unw_is_signal_frame()

libunwind has a function to query whether the cursor points to a signal frame.
Use this to print

 1: <signal handler called>

like GDB does, rather than printing something less useful such as

 1: /usr/lib/libpthread.so.0 (funlockfile+0x60) [0x7f679838b870]

Signed-off-by: Aaron Plattner <aplattner@nvidia.com>
---
 os/backtrace.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/os/backtrace.c b/os/backtrace.c
index 99d776950..bf7ee68e2 100644
--- a/os/backtrace.c
+++ b/os/backtrace.c
@@ -97,9 +97,14 @@ xorg_backtrace(void)
         else
             filename = "?";
 
-        ErrorFSigSafe("%u: %s (%s%s+0x%x) [%p]\n", i++, filename, procname,
-            ret == -UNW_ENOMEM ? "..." : "", (int)off,
-            (void *)(uintptr_t)(ip));
+
+        if (unw_is_signal_frame(&cursor)) {
+            ErrorFSigSafe("%u: <signal handler called>\n", i++);
+        } else {
+            ErrorFSigSafe("%u: %s (%s%s+0x%x) [%p]\n", i++, filename, procname,
+                ret == -UNW_ENOMEM ? "..." : "", (int)off,
+                (void *)(uintptr_t)(ip));
+        }
 
         ret = unw_step(&cursor);
         if (ret < 0)
-- 
2.45.2

